docs/Design_Prototype2.md

```diff
@@ -0,0 +1,756 @@
+    must provide a mechnism to disable this prepending.
+
+* The interface must allow additional metadata in the form of a Hash to
+  be persisted/retrieved with the key-value pair.
+
+#### Backend Plugin Adapter
+
+libkv must provide a backend plugin adapter that
+
+  * loads plugin code provided by libkv and other modules with each catalog
+    compile
+  * instantiates plugins when needed
+  * persists plugins through the lifetime of the catalog compile
+
+    * most efficient for plugins that maintain connections with a
+      key/value service
+
+  * selects and uses the appropriate plugin for each libkv Puppet function call
+    during the catalog compile.
+
+* The plugin adapter must be available to all functions in the Puppet
+  function API.
+* The plugin adapter must be loaded and constructed in a way that prevents
+  cross-environment contamination, when loaded in a puppetserver.
+* The plugin adapter must load plugin software in a way that prevents
+  cross-environment contamination, when loaded in a puppetserver.
+* The plugin adapter must be fault tolerant against any malformed plugin
+  software.
+
+  * It must continue to operate with valid plugins, when a malformed plugin
+    fails to load.
+
+* The plugin adapter must allow multiple instances of an individual
+  plugin to be instantiated and used during the catalog compile, when the
+  instances have different configuration parameters.
+
+#### Backend Plugin API
+
+libkv must supply a backend plugin API that provides
+
+  * Public API method signatures, including the constructor and a
+    method that reports the plugin type (typically backend it supports)
+  * Description of any universal plugin options that must be supported
+  * Ability to specify plugin-specific options
+  * Explicit policy on error handling (how to report errors, what information
+    the messages should contain for plugin proper identification, whether
+    exceptions are allowed)
+  * Details on the code structure required for prevention of cross-environment
+    contamination
+  * Documentation requirements
+  * Testing requirements
+
+Each plugin must conform to the plugin API and satisfy the following
+general requirements:
+
+* All plugins must be unique.
+
+  * Plugin Ruby files can be named the same in different modules, but
+    their reported plugin types must be unique.
+
+* All plugins must allow multiple instances of the plugin to be instantiated
+  and used in a single catalog compile.
+
+  * This requirement allows the same plugin to be used for distinct
+    configurations of the same backend type.
+
+#### Configuration
+
+* Users must be able to specify the following in Hiera:
+
+  * global libkv options
+  * any number of backend configurations
+  * different configurations for the same backend type (e.g., 'file'
+    backend configurations that persist files to different root directories)
+  * the backend configuration to use, always, or a set of defaults
```

Comment:
confusing

Reply:
Default hierarchy needs to be replaced with something simpler.  It was intended to be useful when we thought that libkv could determine the calling resource automatically.  That turned out to not be the case, because function scope is limited unless the function is an internal function.  Although we could make our functions internal, we could not be guaranteed that they would be called within Puppet built-in functions (such as `each`), that would render our attempts to access the caller useless.

I'll rework this backend search algorithm to first look for a full match with the 'resource', if that fails look for a partial (beginning) match, and if that fails, fallback to the 'default' backend.  I'll rename 'resource' to something else for clarity and then fix the docs.